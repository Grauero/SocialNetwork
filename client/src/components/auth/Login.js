import React, { Component } from 'react';
import propTypes from 'prop-types';
import { connect } from 'react-redux';

import { loginUser } from '../../store/actions/authActions';
import TextFieldGroup from '../common/TextFieldGroup';

class Login extends Component {
  state = {
    email: '',
    password: '',
    errors: {}
  };

  componentDidMount() {
    if (this.props.auth.isAuthenticated) {
      this.props.history.push('/dashboard');
    }
  }

  static getDerivedStateFromProps(props, state) {
    if (props.auth.isAuthenticated) {
      props.history.push('/dashboard');
    }

    if (props.errors) {
      return {
        ...state,
        errors: props.errors
      };
    }

    return null;
  }

  onChange = e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  onSubmit = e => {
    e.preventDefault();

    const userData = {
      email: this.state.email,
      password: this.state.password
    };

    this.props.loginUser(userData);
  };

  render() {
    const { email, password, errors } = this.state;

    return (
      <div className="login" style={{ minHeight: '100vh', paddingTop: '7em' }}>
        <div className="container">
          <div className="row">
            <div className="col-md-8 m-auto">
              <h1 className="display-4 text-center">Log In</h1>
              <p className="lead text-center">
                Sign in to your DevNetwork account
              </p>
              <form action="dashboard.html" onSubmit={this.onSubmit}>
                <TextFieldGroup
                  placeholder="Email Address"
                  name="email"
                  type="email"
                  value={email}
                  onChange={this.onChange}
                  error={errors.email}
                />
                <TextFieldGroup
                  placeholder="Password"
                  name="password"
                  type="password"
                  value={password}
                  onChange={this.onChange}
                  error={errors.password}
                />
                <input type="submit" className="btn btn-info btn-block mt-4" />
              </form>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  auth: propTypes.instanceOf(Object).isRequired,
  errors: propTypes.instanceOf(Object).isRequired,
  history: propTypes.instanceOf(Object).isRequired,
  loginUser: propTypes.func.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors
});

export { Login };
export default connect(
  mapStateToProps,
  { loginUser }
)(Login);
