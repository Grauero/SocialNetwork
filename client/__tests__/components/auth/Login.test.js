import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { Login } from '../../../src/components/auth/Login';

const props = {
  auth: { isAuthenticated: false },
  errors: {},
  history: { push: jest.fn() },
  loginUser: jest.fn()
};
const initialState = {
  email: '',
  password: '',
  errors: {}
};

const wrapper = mount(<Login {...props} />);

const component = wrapper.find(Login);
component.setState(initialState);

it('should redirect to /dashboard if user is authenticated', () => {
  const authProps = Object.assign({}, props, {
    auth: { isAuthenticated: true }
  });
  mount(<Login {...authProps} />);

  expect(authProps.history.push).toHaveBeenCalled();
  expect(authProps.history.push).toHaveBeenCalledWith('/dashboard');
});

it('handles form submit', () => {
  const expectedObj = { email: '', password: '' };
  component.find('form').simulate('submit');

  expect(props.loginUser).toHaveBeenCalled();
  expect(props.loginUser).toHaveBeenCalledWith(expectedObj);
  expect(props.loginUser).toHaveBeenCalledTimes(1);
});

it('handles inputs change', () => {
  const name = 'email';
  const value = 'updatedValue';
  const expectedState = Object.assign({}, initialState);
  expectedState[name] = value;

  component
    .find('TextFieldGroup')
    .at(0)
    .find('input')
    .simulate('change', {
      target: { name, value }
    });

  expect(component.state()).toEqual(expectedState);
});

it('matches snapshot', () => {
  expect(toJSON(component)).toMatchSnapshot();
});
