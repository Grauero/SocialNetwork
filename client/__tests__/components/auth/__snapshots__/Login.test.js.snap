// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Login
  auth={
    Object {
      "isAuthenticated": false,
    }
  }
  errors={Object {}}
  history={
    Object {
      "push": [Function],
    }
  }
  loginUser={[Function]}
>
  <div
    className="login"
    style={
      Object {
        "minHeight": "100vh",
        "paddingTop": "7em",
      }
    }
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-md-8 m-auto"
        >
          <h1
            className="display-4 text-center"
          >
            Log In
          </h1>
          <p
            className="lead text-center"
          >
            Sign in to your DevConnector account
          </p>
          <form
            action="dashboard.html"
            onSubmit={[Function]}
          >
            <TextFieldGroup
              disabled={null}
              error={null}
              info={null}
              name="email"
              onChange={[Function]}
              placeholder="Email Address"
              type="email"
              value=""
            >
              <div
                className="form-group"
              >
                <input
                  className="form-control form-control-lg"
                  disabled={null}
                  name="email"
                  onChange={[Function]}
                  placeholder="Email Address"
                  type="email"
                  value=""
                />
              </div>
            </TextFieldGroup>
            <TextFieldGroup
              disabled={null}
              error={null}
              info={null}
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
            >
              <div
                className="form-group"
              >
                <input
                  className="form-control form-control-lg"
                  disabled={null}
                  name="password"
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
            </TextFieldGroup>
            <input
              className="btn btn-info btn-block mt-4"
              type="submit"
            />
          </form>
        </div>
      </div>
    </div>
  </div>
</Login>
`;
